File: test/src/toml.js
import test from 'oro:test'
import {
  parse,
  stringify,
  TomlLocalDate,
  TomlLocalTime,
  TomlLocalDateTime
} from 'oro:toml'

test('toml.parse basic scalars', (t) => {
  const doc = `
title = "Example"
enabled = true
count = 42
ratio = 3.1415
`
  const cfg = parse(doc)
  t.equal(cfg.title, 'Example', 'parses strings')
  t.equal(cfg.enabled, true, 'parses booleans')
  t.equal(cfg.count, 42, 'parses integers')
  t.equal(cfg.ratio, 3.1415, 'parses floats')
})

test('toml.parse handles string variants', (t) => {
  const doc = `
basic = "Line 1\\nLine 2"
literal = 'C:\\\\Path\\\\file.txt'
multiline = """
Roses are red\\
  Violets are blue
"""
literal_multiline = '''
Line A
Line B "quoted"
'''
`
  const cfg = parse(doc)
  t.equal(cfg.basic, 'Line 1\nLine 2', 'supports escaped newline')
  t.equal(cfg.literal, 'C:\\Path\\file.txt', 'supports literal strings')
  t.equal(
    cfg.multiline,
    'Roses are redViolets are blue',
    'supports line continuations'
  )
  t.equal(
    cfg.literal_multiline,
    'Line A\nLine B "quoted"\n',
    'supports literal multiline strings'
  )
})

test('toml.parse rejects unterminated strings', (t) => {
  t.throws(
    () => parse('title = "abc'),
    /unterminated basic string/i,
    'detects unterminated basic string'
  )
  t.throws(
    () => parse("path = 'abc"),
    /unterminated literal string/i,
    'detects unterminated literal string'
  )
  t.throws(
    () => parse('note = """abc'),
    /unterminated multiline basic string/i,
    'detects unterminated multiline basic string'
  )
  t.throws(
    () => parse("msg = '''abc"),
    /unterminated multiline literal string/i,
    'detects unterminated multiline literal string'
  )
})

test('toml.parse numbers and special floats', (t) => {
  const doc = `
hex = 0xDEADBEEF
oct = 0o755
bin = 0b1111_0000
nan = nan
inf = +inf
`
  const cfg = parse(doc)
  t.equal(cfg.hex, 0xdeadbeef, 'parses hexadecimal integer')
  t.equal(cfg.oct, 0o755, 'parses octal integer')
  t.equal(cfg.bin, 0b11110000, 'parses binary integer')
  t.ok(Number.isNaN(cfg.nan), 'parses NaN')
  t.equal(cfg.inf, Infinity, 'parses Infinity')
})

test('toml.parse enforces 64-bit integer range', (t) => {
  t.throws(
    () => parse('value = 9223372036854775808'),
    /out of range/i,
    'rejects integers above int64 max'
  )
  t.throws(
    () => parse('value = -9223372036854775809'),
    /out of range/i,
    'rejects integers below int64 min'
  )
  t.throws(
    () => parse('value = 0x8000000000000000'),
    /out of range/i,
    'rejects oversized hexadecimal integers'
  )
  const within = parse('value = 9223372036854775807')
  t.ok(
    typeof within.value === 'number' && Number.isFinite(within.value),
    'allows int64 max value'
  )
})

test('toml.parse rejects leading zero floats', (t) => {
  t.throws(
    () => parse('value = 01.0'),
    /Invalid numeric value/,
    'rejects leading zero before decimal'
  )
  t.throws(
    () => parse('value = 01e2'),
    /Invalid numeric value/,
    'rejects leading zero before exponent'
  )
})

test('toml.parse date and time types', (t) => {
  const doc = `
offset = 1979-05-27T07:32:00Z
local_dt = 1979-05-27T00:32:00
local_date = 1979-05-27
local_time = 07:32:00.123456
`
  const cfg = parse(doc)
  t.ok(cfg.offset instanceof Date, 'offset date-time becomes Date')
  t.equal(
    cfg.offset.toISOString(),
    '1979-05-27T07:32:00.000Z',
    'offset date-time keeps zone'
  )
  t.ok(
    cfg.local_dt instanceof TomlLocalDateTime,
    'local date-time returns TomlLocalDateTime'
  )
  t.ok(cfg.local_dt.date instanceof TomlLocalDate, 'local date captured')
  t.ok(cfg.local_dt.time instanceof TomlLocalTime, 'local time captured')
  t.ok(cfg.local_date instanceof TomlLocalDate, 'local date parsed')
  t.ok(cfg.local_time instanceof TomlLocalTime, 'local time parsed')
  t.equal(
    cfg.local_time.nanosecond,
    123456000,
    'precision preserved for fractional seconds'
  )
})

test('toml.parse rejects invalid offset date-times', (t) => {
  t.throws(
    () => parse('ts = 1979-02-31T07:32:00Z'),
    /invalid offset date-time/i,
    'rejects impossible calendar dates'
  )
  t.throws(
    () => parse('ts = 1979-05-27T07:32:00+24:00'),
    /invalid offset date-time/i,
    'rejects invalid offset hours'
  )
})

test('toml.parse tables and arrays of tables', (t) => {
  const doc = `
[database]
ports = [ 8000, 8001, 8002 ]
connection_max = 5000

[servers.alpha]
ip = "10.0.0.1"

[servers.beta]
ip = "10.0.0.2"

[[products]]
name = "Hammer"
sku = 738594937

[[products]]
name = "Nail"
sku = 284758393
`
  const cfg = parse(doc)
  t.deepEqual(
    cfg.database.ports,
    [8000, 8001, 8002],
    'nested table array parsed'
  )
  t.equal(cfg.servers.alpha.ip, '10.0.0.1', 'nested table alpha parsed')
  t.equal(cfg.servers.beta.ip, '10.0.0.2', 'nested table beta parsed')
  t.equal(cfg.products.length, 2, 'array of tables yields array of objects')
  t.equal(cfg.products[0].name, 'Hammer', 'first product name parsed')
  t.equal(cfg.products[1].sku, 284758393, 'second product sku parsed')
})

test('toml.parse dotted keys and inline tables', (t) => {
  const doc = `
user.name = "alice"
user.contact.email = "alice@example.com"
settings = { theme = "dark", window = { width = 800, height = 600 } }
`
  const cfg = parse(doc)
  t.equal(cfg.user.name, 'alice', 'dotted key created nested table')
  t.equal(cfg.user.contact.email, 'alice@example.com', 'deep dotted key works')
  t.equal(cfg.settings.theme, 'dark', 'inline table parsed')
  t.equal(cfg.settings.window.width, 800, 'inline nested table parsed')
})

test('toml.parse enforces array homogeneity', (t) => {
  t.throws(
    () => parse('mixed = [1, "two"]'),
    /homogeneous/i,
    'mixed arrays rejected'
  )
})

test('toml.parse prevents modifying inline tables', (t) => {
  t.throws(
    () => parse('config = { key = 1 }\nconfig.other = 2'),
    /inline table/i,
    'inline table is sealed'
  )
})

test('toml.parse prevents redefining tables via headers', (t) => {
  const doc = `
[app]
name = "one"

[app]
name = "two"
`
  t.throws(
    () => parse(doc),
    /already defined/i,
    'duplicate table header rejected'
  )
})

test('toml.parse prevents redefining implicit tables', (t) => {
  const doc = `
app.window.width = 800

[app.window]
height = 600
`
  t.throws(
    () => parse(doc),
    /already defined/i,
    'implicit table cannot be redefined'
  )
})

test('toml.parse supports reviver', (t) => {
  const doc = 'value = 2'
  const cfg = parse(doc, {
    reviver (key, value) {
      if (key === 'value') return value * 10
      return value
    }
  })
  t.equal(cfg.value, 20, 'reviver transforms values')
})

test('toml.stringify basic object', (t) => {
  const data = {
    title: 'Example',
    enabled: true,
    count: 42,
    ratio: 3.1415,
    tags: ['alpha', 'beta']
  }
  const doc = stringify(data)
  const roundTrip = parse(doc)
  t.deepEqual(roundTrip, data, 'stringify round-trips simple types')
})

test('toml.stringify nested tables and arrays of tables', (t) => {
  const data = {
    database: {
      server: '192.168.1.1',
      ports: [8000, 8001],
      connection_max: 5000
    },
    servers: {
      alpha: { ip: '10.0.0.1' },
      beta: { ip: '10.0.0.2' }
    },
    products: [
      { name: 'Hammer', sku: 738594937 },
      { name: 'Nail', sku: 284758393 }
    ]
  }
  const doc = stringify(data)
  const roundTrip = parse(doc)
  t.deepEqual(roundTrip, data, 'nested tables preserve structure')
  t.match(
    doc,
    /\[\[products\]\]\nname = "Hammer"\n/,
    'array of tables emitted with headers'
  )
})

test('toml.stringify temporal values', (t) => {
  const data = {
    offset: new Date('1979-05-27T07:32:00Z'),
    localDate: new TomlLocalDate(1979, 5, 27),
    localTime: new TomlLocalTime(7, 32, 0, 123000000),
    localDateTime: new TomlLocalDateTime(
      new TomlLocalDate(1979, 5, 27),
      new TomlLocalTime(0, 32, 0, 0)
    )
  }
  const doc = stringify(data)
  const roundTrip = parse(doc)
  t.equal(
    roundTrip.offset.toISOString(),
    data.offset.toISOString(),
    'offset datetime round-trips through Date ISO'
  )
  t.equal(
    roundTrip.localDate.toString(),
    data.localDate.toString(),
    'local date preserved'
  )
  t.equal(
    roundTrip.localTime.toString(),
    data.localTime.toString(),
    'local time preserved'
  )
  t.equal(
    roundTrip.localDateTime.toString(),
    data.localDateTime.toString(),
    'local date-time preserved'
  )
})

test('toml.stringify rejects mixed arrays', (t) => {
  const data = { mixed: [1, 'two'] }
  t.throws(() => stringify(data), /single type/i, 'mixed arrays are rejected')
})

File: test/src/runtime-core/toml.cc
#include "tests.hh"
#include "src/runtime/toml.hh"

namespace oro::Tests {
  void toml (Harness& t) {
    t.test("TOML basic values", [] (auto t) {
      const auto document = oro::runtime::TOML::parse(R"TOML(
        title = "TOML Example"
        enabled = true
        answer = 42
        ratio = 1.5
      )TOML");

      const auto& table = document.asTable();

      t.equals(table.at("title").asString(), "TOML Example", "title parsed");
      t.equals(table.at("enabled").asBool(), true, "boolean parsed");
      t.equals(table.at("answer").asInteger(), 42, "integer parsed");
      t.assert(table.at("ratio").asFloat() == 1.5, "float parsed");
    });

    t.test("TOML tables and arrays", [] (auto t) {
      const auto document = oro::runtime::TOML::parse(R"TOML(
        [database]
        server = "192.168.1.1"
        ports = [
          8001,
          8001, # repeated port
          8002
        ]

        trailing = [
          "one",
          "two",
        ]

        inline_trailing = [1, 2,]

        [[clients]]
        name = "alpha"
        [clients.preferences]
        color = "blue"

        [[clients]]
        name = "beta"
      )TOML");

      const auto& root = document.asTable();
      const auto& database = root.at("database").asTable();
      t.equals(database.at("server").asString(), "192.168.1.1", "server parsed");

      const auto& ports = database.at("ports").asArray();
      t.equals(static_cast<int64_t>(ports.size()), 3, "ports array size");
      t.equals(ports[2].asInteger(), 8002, "ports third entry");

      const auto& clients = root.at("clients").asArray();
      t.equals(static_cast<int64_t>(clients.size()), 2, "clients array size");

      const auto& firstClient = clients[0].asTable();
      t.equals(firstClient.at("name").asString(), "alpha", "first client name");
      const auto& preferences = firstClient.at("preferences").asTable();
      t.equals(preferences.at("color").asString(), "blue", "nested table inherits array context");

      const auto& secondClient = clients[1].asTable();
      t.equals(secondClient.at("name").asString(), "beta", "second client name");

      const auto& trailing = database.at("trailing").asArray();
      t.equals(static_cast<int64_t>(trailing.size()), 2, "trailing comma array size");
      t.equals(trailing[1].asString(), "two", "trailing comma retains value");

      const auto& inlineTrailing = database.at("inline_trailing").asArray();
      t.equals(static_cast<int64_t>(inlineTrailing.size()), 2, "inline trailing comma array size");
      t.equals(inlineTrailing[0].asInteger(), 1, "inline trailing first value");
      t.equals(inlineTrailing[1].asInteger(), 2, "inline trailing second value");
    });

    t.test("TOML date and time types", [] (auto t) {
      const auto document = oro::runtime::TOML::parse(R"TOML(
        local_date = 1979-05-27
        local_time = 07:32:00
        local_datetime = 1979-05-27T07:32:00
        offset_datetime = 1979-05-27T07:32:00+01:00
      )TOML");

      const auto& table = document.asTable();

      const auto& date = table.at("local_date").asDate();
      t.equals(static_cast<int64_t>(date.year), 1979, "date year");
      t.equals(static_cast<int64_t>(date.month), 5, "date month");
      t.equals(static_cast<int64_t>(date.day), 27, "date day");

      const auto& time = table.at("local_time").asTime();
      t.equals(static_cast<int64_t>(time.hour), 7, "time hour");
      t.equals(static_cast<int64_t>(time.minute), 32, "time minute");
      t.equals(static_cast<int64_t>(time.second), 0, "time second");

      const auto& localDateTime = table.at("local_datetime").asDateTime();
      t.equals(static_cast<int64_t>(localDateTime.date.year), 1979, "local datetime year");
      t.equals(localDateTime.hasOffset, false, "local datetime has no offset");

      const auto& offsetDateTime = table.at("offset_datetime").asDateTime();
      t.equals(static_cast<int64_t>(offsetDateTime.offsetMinutes), 60, "offset datetime offset");
      t.equals(offsetDateTime.hasOffset, true, "offset datetime has offset");
    });

    t.test("TOML multiline and inline tables", [] (auto t) {
      const auto document = oro::runtime::TOML::parse(R"TOML(
        poem = """Roses are red
Violets are blue"""
        inline = { nested = { value = 1 }, label = "test" }
      )TOML");

      const auto& table = document.asTable();
      t.equals(table.at("poem").asString(), "Roses are red\nViolets are blue", "multiline string preserves newlines");

      const auto& inlineTable = table.at("inline").asTable();
      t.equals(inlineTable.at("label").asString(), "test", "inline table string");
      const auto& nested = inlineTable.at("nested").asTable();
      t.equals(nested.at("value").asInteger(), 1, "inline nested value");
    });

    t.test("TOML inline table immutability", [] (auto t) {
      t.throws([] () {
        oro::runtime::TOML::parse(R"TOML(
          inline = { nested = 1 }
          inline.nested = 2
        )TOML");
      }, "inline tables cannot be extended with dotted keys");
    });

    t.test("TOML dotted keys conflict with later table headers", [] (auto t) {
      t.throws([] () {
        oro::runtime::TOML::parse(R"TOML(
          app.window.width = 800

          [app.window]
          height = 600
        )TOML");
      }, "implicit tables cannot be redefined by headers");
    });
  }
}
